### PR TITLE
Chore: Move StreamingDataFrame to @grafana/data

### LINKED ISSUE #71329: Remove references to Grafana internal code in the test data source
Part of https://github.com/grafana/grafana/issues/68973

### PR DESCRIPTION
**Pull request by andresmgot**
**What is this feature?**

The `StreamingDataFrame` class is used by multiple core plugins (test data, prometheus, loki). As part of the effort in https://github.com/grafana/grafana/issues/71329, this PR moves the class (and its dependency `renderLegendFormat`) to @grafana/data to remove the privileged import from these core plugins.

**Which issue(s) does this PR fix?**:

Closes https://github.com/grafana/grafana/issues/71329

**Special notes for your reviewer:**

Please check that:
- [X] It works as expected from a user's perspective.
- [X] If this is a pre-GA feature, it is behind a feature toggle.
- [X] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: packages/grafana-data/src/dataframe/StreamingDataFrame.ts [hunk 1]
```diff
@@ -1,21 +1,38 @@
...
-} from '@grafana/data';
-import { join } from '@grafana/data/src/transformations/transformers/joinDataFrames';
-import { StreamingFrameAction, StreamingFrameOptions } from '@grafana/runtime/src/services/live';
-import { renderLegendFormat } from 'app/plugins/datasource/prometheus/legend';
-import { AlignedData } from 'uplot';
+} from '.';
+
+/**
+ * Indicate if the frame is appened or replace
+ * From @grafana/runtime/src/services/live
```
Comment (andresmgot) [inline packages/grafana-data/src/dataframe/StreamingDataFrame.ts:16]: I am copying these (simple) types from @grafana/runtime. I found no better option but maybe I am missing something?

File: packages/grafana-data/src/dataframe/index.ts [hunk 2]
```diff
@@ -8,3 +8,9 @@ export * from './ArrayDataFrame';
 export * from './DataFrameJSON';
 export * from './frameComparisons';
 export { anySeriesWithTimeField, isTimeSeriesFrame, isTimeSeriesFrames } from './utils';
+export {
+  StreamingDataFrame,
+  closestIdx,
+  getStreamingFrameOptions,
+  getLastStreamingDataFramePacket,
```
Comment (ryantxu) [inline packages/grafana-data/src/dataframe/index.ts:15]:
```suggestion
```
I don't think we need to expose these publicly yet.  We can internally import them with the full path

File: packages/grafana-data/src/dataframe/StreamingDataFrame.ts (hunk 1 above)
Comment (ryantxu) [reply packages/grafana-data/src/dataframe/StreamingDataFrame.ts:16]: rather than copy -- can we "move" them to data.  essentially export here, then re-export in grafana data

File: packages/grafana-data/src/dataframe/StreamingDataFrame.ts [hunk 3]
```diff
@@ -98,6 +115,7 @@ export class StreamingDataFrame implements DataFrame {
     const dataFrameDTO = toFilteredDataFrameDTO(this, fieldPredicate);
 
     const numberOfItemsToRemove = getNumberOfItemsToRemove(
+      // eslint-disable-next-line
```
Comment (ashharrison90) [inline packages/grafana-data/src/dataframe/StreamingDataFrame.ts:118]: have fixed the problem with betterer - you can remove all these `eslint-disable-next-line` comments now 👍

File: public/app/features/live/live.test.ts [hunk 4]
```diff
@@ -1,10 +1,10 @@
 import { Subject } from 'rxjs';
 
 import { DataQueryResponse, FieldType, LiveChannelScope } from '@grafana/data';
+import { StreamingDataFrame } from '@grafana/data/src/dataframe/StreamingDataFrame';
```
Comment (ryantxu) [inline public/app/features/live/live.test.ts:4]:
```suggestion
import { DataQueryResponse, FieldType, LiveChannelScope, StreamingDataFrame } from '@grafana/data';
```

**Review by ryantxu — APPROVED** [review]
LGTM -- thanks @andresmgot

File: public/app/features/live/centrifuge/LiveDataStream.ts [hunk 5]
```diff
@@ -10,11 +10,12 @@ import {
   LiveChannelEvent,
   LiveChannelId,
   LoadingState,
+  StreamingDataFrame,
 } from '@grafana/data';
+import { getStreamingFrameOptions } from '@grafana/data/src/dataframe/StreamingDataFrame';
```
Comment (mckn) [inline public/app/features/live/centrifuge/LiveDataStream.ts:15]: nit: import from `@grafana/data` instead.

File: public/app/features/live/data/amendTimeSeries.ts [hunk 6]
```diff
@@ -1,4 +1,4 @@
-import { closestIdx } from "./StreamingDataFrame";
+import { closestIdx } from '@grafana/data/src/dataframe/StreamingDataFrame';
```
Comment (mckn) [inline public/app/features/live/data/amendTimeSeries.ts:1]: nit: import from `@grafana/data` instead.

File: public/app/plugins/panel/state-timeline/StateTimelinePanel.tsx [hunk 7]
```diff
@@ -1,6 +1,7 @@
 import React, { useCallback, useMemo, useRef, useState } from 'react';
 
 import { CartesianCoords2D, DashboardCursorSync, DataFrame, FieldType, PanelProps } from '@grafana/data';
+import { getLastStreamingDataFramePacket } from '@grafana/data/src/dataframe/StreamingDataFrame';
```
Comment (mckn) [inline public/app/plugins/panel/state-timeline/StateTimelinePanel.tsx:4]: nit: import from `@grafana/data` instead?

File: .betterer.results [hunk 8]
```diff
@@ -55,6 +55,16 @@ exports[`better eslint`] = {
...
+    "packages/grafana-data/src/dataframe/StreamingDataFrame.ts:5381": [
+      [0, 0, 0, "Do not use any type assertions.", "0"],
+      [0, 0, 0, "Do not use any type assertions.", "1"],
+      [0, 0, 0, "Do not use any type assertions.", "2"],
+      [0, 0, 0, "Do not use any type assertions.", "3"],
+      [0, 0, 0, "Do not use any type assertions.", "4"],
+      [0, 0, 0, "Unexpected any. Specify a different type.", "5"],
+      [0, 0, 0, "Do not use any type assertions.", "6"],
+      [0, 0, 0, "Do not use any type assertions.", "7"]
+    ],
```
Comment (mckn) [inline .betterer.results:67]: We can probably get rid of a lot of these but lets do that in a separate PR.

**Review by mckn — COMMENTED** [review]
LGTM! Added some small nits.

File: public/app/features/live/data/amendTimeSeries.ts (hunk 6 above)
Comment (andresmgot) [reply public/app/features/live/data/amendTimeSeries.ts:1]: same, it's not exported either

File: public/app/plugins/panel/state-timeline/StateTimelinePanel.tsx (hunk 7 above)
Comment (andresmgot) [reply public/app/plugins/panel/state-timeline/StateTimelinePanel.tsx:4]: same, not exported

**Review by mckn — APPROVED** [review]
LGTM!
